presto-main/src/main/java/com/facebook/presto/operator/NestedLoopJoinOperator.java

```diff
@@ -0,0 +1,237 @@
+
+public class NestedLoopJoinOperator
+        implements Operator, Closeable
+{
+    private final ListenableFuture<LookupSource> lookupSourceFuture;
+
+    private final OperatorContext operatorContext;
+    private final List<Type> probeTypes;
+    private final List<Type> types;
+    private final PageBuilder pageBuilder;
+    private final int numProbeColumns;
+
+    private PagesIndexLookupSource lookupSource;
+    private Page probePage;
+
+    private boolean finishing;
+    private int probePosition = -1;
+    private int lookupPosition = -1;
+
+    private List<Object> probeRow = null;
+    private List<List<Object>> lookupRows = null;
+
+    public NestedLoopJoinOperator(
+            OperatorContext operatorContext,
+            LookupSourceSupplier lookupSourceSupplier,
+            List<Type> probeTypes)
+    {
+        this.operatorContext = checkNotNull(operatorContext, "operatorContext is null");
+        this.lookupSourceFuture = checkNotNull(lookupSourceSupplier, "lookupSourceSupplier is null").getLookupSource(operatorContext);
+        this.probeTypes = checkNotNull(probeTypes, "probeTypes is null");
+        this.numProbeColumns = probeTypes.size();
+
+        this.types = ImmutableList.<Type>builder()
+                .addAll(probeTypes)
+                .addAll(lookupSourceSupplier.getTypes())
+                .build();
+        this.pageBuilder = new PageBuilder(types);
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    @Override
+    public void finish()
+    {
+        finishing = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        boolean finished = finishing && probePage == null && pageBuilder.isEmpty();
+
+        // if finished drop references so memory is freed early
+        if (finished) {
+            lookupSource = null;
+            probePage = null;
+            pageBuilder.reset();
+        }
+        return finished;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        return lookupSourceFuture;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        if (finishing || probePage != null) {
+            return false;
+        }
+
+        if (lookupSource == null) {
+            lookupSource = (PagesIndexLookupSource) tryGetUnchecked(lookupSourceFuture);
+        }
+        return lookupSource != null;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        checkNotNull(page, "page is null");
+        checkState(!finishing, "Operator is finishing");
+        checkState(lookupSource != null, "Lookup source has not been built yet");
+        checkState(probePage == null, "Current page has not been completely processed yet");
+
+        probePage = page;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (lookupSource == null) {
```

Comment:
you don't need this check.  If you have a probe page, you are guaranteed to have a lookup source.